### DISCUSSION TITLE
Best approach to applying values loaded asychronously?

### DISCUSSION THREAD

**Discussion by raulfelix** [react-hook-form/react-hook-form#1696] [General] [answered]
Say I have a form and the values for the form will be retrieved from an API.

What is the best approach to setting the form values? Should we use `setValue`?

Here is an example which I think used to work in previous versions. The name field is set in the form as a `defaultValue` but after it is loaded the field is not refreshed to reflect the loaded value.

https://codesandbox.io/s/empty-tree-qff1x?file=/src/App.js

**Comment by bluebill1049** [comment, marked as answer]
hey @raulfelix, it's better to use `reset()` API


```
useEffect(() => {
  reset(values);
}, [values])
```

**Reply by raulfelix** [reply]
Thank you!

**Reply by mitchconquer** [reply]
Hi, I have a follow up to this. If you use `reset` it seems to delete all of the default values and there's no way to set them again so if the user navigates away from the form fields, their new values that I was setting with `defaultValue` are not shown. Is there a way to reset without blowing away the default values?

**Reply by mitchconquer** [reply]
I believe it's [this line](https://github.com/react-hook-form/react-hook-form/blob/a5709440eb2e6a908110aea9ed473772ded5686a/src/useForm.ts#L1132). I wonder if it would be useful to others as well to have it behind a conditional check so that reseting default values is optional?

**Reply by bluebill1049** [reply]
what about this?
```
reset({
  ...getValues(),
  ....yourData,
})
```

**Reply by mitchconquer** [reply]
I have tried that but the problem is that if the user makes a change and navigates away and back to the form, the updated values from after the reset are not shown because I rely on the defaultValue property to display field updates. Maybe this is not a correct use of RHF.

**Reply by bluebill1049** [reply]
is the form gets unmounted when user navigate away?

**Reply by mitchconquer** [reply]
The inputs do, yes.

**Reply by bluebill1049** [reply]
that's the problem, you will need to store those values in a state lib. so they don't get unregistered when unmount.

**Reply by mitchconquer** [reply]
Ok, I see. I noticed it worked if I didn't reset first (ie: set the values by iterating through and using `setValue`) but reset was preferable because I wanted to get the validation result before I rendered the form.

But I think it's stretching the library beyond its intended use case so I will use state for this form.

Thanks so much for the support!

**Comment by kkfuri** [comment]
Is it safe to use the `reset()` API in the useEffect dependency list, following the react-hooks/exhaustive-deps?
```
useEffect(() => {
    reset(values);
  }, [values, reset]);
```

**Reply by bluebill1049** [reply]
yes reset is memoed inside hookform

**Reply by kkfuri** [reply]
Thank you! 😁

**Comment by caiorios** [comment]
Can I use `reset()` multiple times inside a form? In my case I keep my form components separate, each one of them having it's own logic for async load.

**Reply by bluebill1049** [reply]
yes :)

**Reply by caiorios** [reply]
Thank you!
